File: iOS/LeetCodeSwift/LeetCodeSwift/Leetcode/Valid_Parentheses.swift
//
//  Valid_Parentheses.swift
//  LeetCodeSwift
//
//  Created by tree on 2018/3/22.
//  Copyright © 2018年 treee. All rights reserved.
//

import Foundation
extension Solution {
    /**
     Question:
         Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
     
         The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
     */
    /// 见代码
    func isValid(_ s: String) -> Bool {
        var stack: [String] = []
        let pushSyn = ["(", "[", "{"]
        for c in s {
            let sub = String.init(c)
            if pushSyn.contains(sub) {
                stack.append(sub)
            }
            else if sub == ")" && stack.last == "(" && stack.isEmpty == false {
                stack.removeLast()
            }
            else if sub == "]" && stack.last == "[" && stack.isEmpty == false {
                stack.removeLast()
            }
            else if sub == "}" && stack.last == "{" && stack.isEmpty == false {
                stack.removeLast()
            }else { return false }
        }
        return stack.isEmpty
    }
}

File: python/LeetCode/Valid_Parentheses.py


'''
有效的括号

给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串，判断字符串是否有效。

括号必须以正确的顺序关闭，"()" 和 "()[]{}" 是有效的但是 "(]" 和 "([)]" 不是。

'''
class Solution(object):
    def isValid(self, s: str) -> bool:
        stack = []
        info = {')': '(',
                ']': '[',
                '}': '{'}
        for char in s:
            if char in info.values():
                stack.append(char)
            elif char in info.keys():
                if stack == [] or info[char] != stack.pop():
                    return False
            else:
                return False
        return stack == []
